packages/compass-import-export/src/modules/export.ts

```diff
@@ -1,676 +1,237 @@
-    try {
-      count =
-        maybeCount ??
-        (!isAggregation
-          ? await fetchDocumentCount(dataService!, namespace, query)
-          : null);
-    } catch (e: unknown) {
-      dispatch(onError(e as Error));
-    }
-    dispatch(nsChanged(namespace));
-    dispatch(onModalOpen({ namespace, query, count, aggregation }));
+import type { SchemaPath } from '../export/gather-fields';
+import type {
+  ExportAggregation,
+  ExportQuery,
+  ExportResult,
+} from '../export/export-types';
+import { queryHasProjection } from '../utils/query-has-projection';
+
+// TODO(COMPASS-6426): Update the fields to export type based on the gather fields result type.
+export type FieldsToExport = {
+  [fieldId: string]: {
+    fieldPath: SchemaPath;
+    selected: boolean;
   };
-
-const getQuery = (query: ExportQueryType | null, isFullCollection: boolean) => {
-  if (isFullCollection || !query) {
-    return FULL_QUERY;
-  }
-  return query;
 };
 
-export const sampleFields =
-  (): ThunkAction<void, RootExportState, void, AnyAction> =>
-  async (
-    dispatch: ThunkDispatch<RootExportState, void, AnyAction>,
-    getState: () => RootExportState
-  ) => {
+// Fields can only be prefixed with one '$'. Otherwise $
+// is not allowed, so this should be a safe separator.
+const FieldSplittingSymbol = '$$$$';
+function getIdForSchemaPath(schemaPath: SchemaPath) {
+  return schemaPath.join(FieldSplittingSymbol);
+}
+export function getLabelForFieldId(fieldId: string) {
+  return fieldId.split(FieldSplittingSymbol).join('.');
+}
+
+export const selectFieldsToExport = createAsyncThunk<
+  FieldsToExport,
+  void,
+  { state: RootExportState }
+>(
+  'export/select-fields-to-export',
+  async (_, { getState, rejectWithValue }): Promise<FieldsToExport> => {
     const {
-      ns,
-      exportData,
+      export: { query, namespace, fieldsToExportAbortController },
       dataService: { dataService },
     } = getState();
 
-    const spec = getQuery(exportData.query, exportData.isFullCollection);
-
-    try {
-      const allFields = await loadFields(dataService!, ns, {
-        filter: spec.filter,
-        sampleSize: 50,
-      });
-      const selectedFields = getSelectableFields(allFields, {
-        maxDepth: 2,
-      });
-
-      dispatch(updateAllFields(allFields));
-      dispatch(updateSelectedFields(selectedFields));
-    } catch (err) {
-      // ignoring the error here so users can still insert
-      // fields manually
-      debug('failed to load fields', err);
+    await new Promise((resolve) => setTimeout(resolve, 1000));
+
+    // TODO(COMPASS-6426): Update the interface of this method, pass the whole query.
+    const schemaPaths = await gatherFields({
+      ns: namespace,
+      abortSignal: fieldsToExportAbortController?.signal,
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      dataService: dataService!,
+      filter: query?.filter || {},
+      sampleSize: 10, // TODO: Decide default size.
```

Comment:
@lerouxb Any thoughts on default sample size? We could do have a UX with a skip/stop button, but I think that would be more annoying to folks. 10 seems to little. Currently Compass uses 50, should we update to that? https://github.com/mongodb-js/compass/blob/f2a73a756a616db9e38908d43436c7a52c31ac4f/packages/compass-import-export/src/modules/export.ts#L438

Reply:
50 is fine. 1000 would probably have been fine as well, but I suppose users could have multi-megabyte documents. Previously we loaded them all into memory at once and in the new gatherFields() we're streaming so it shouldn't matter too much.

I think just do 50 and then as a stretch goal (something we discussed in design sync ages ago) we can add a button to load everything that will have the skip button. So you quickly get something out and then if you want you can start loading everything. And then still change your mind anyway.

I'm thinking for the edge case where the sample misses a lot of fields that you would have to add using the form one by one it might be advantageous to have the opt-in skippable one (because it could still be faster and a better experience than adding them all manually), but anything short of that 50 is probably fine.

There's also a point where skipping this whole step and just exporting everything starts to make sense. Or just writing a projection manually. So we probably don't have to overthink it too much.